### ISSUE TITLE
Custom macro parsing in special use cases

### ISSUE THREAD

**Issue by agausmann** [rust-lang/rust-bindgen#1754]
In working on `emscripten-sys`, I have a scenario where `#define` macros are being used as type aliases, which I would like to translate to Rust.

Since I don't think this is a very common pattern, if you don't wish to merge this upstream, I think it could be resolved by somehow allowing the callee to specify custom behavior via `ParseCallbacks` for macros that `bindgen` doesn't know how to handle.

### Input C/C++ Header

types.h:

```C++
#define EM_RESULT int
```

### Bindgen Invocation

```Rust
fn main() {
    let builder = bindgen::builder()
        .header("types.h");

    builder
        .generate()
        .expect("failed to generate rust bindings")
        .write(Box::new(std::io::stdout()))
        .expect("failed to write to file");
}
```

### Actual Results

No bindings are generated.

### Expected Results

```rust
pub type EM_RESULT = c_int;
```

**Comment by emilio** [comment]
Yeah the problem with doing this automatically is that you literally can't. C macros are pretty bad at this kind of thing... Maybe we should add a macro which would generate arbitrary code? But then I guess that users of that macro would use `c_int` instead of `EMSCRIPTEN_RESULT`...

**Comment by agausmann** [comment]
Yes, for arbitrary types it's pretty much impossible, but if it were limited to primitives (those in `std::os::raw`, plus pointers maybe), it would be possible.

**Comment by emilio** [comment]
Even for simple types, you don't really know what the type is without actually expanding the macro.
